packages/core-data/src/resolvers.js

```diff
@@ -107,3 +111,22 @@ export async function* getEmbedPreview( state, url ) {
 		yield receiveEmbedPreview( url, false );
 	}
 }
+
+/**
+ * Requests Upload Permissions from the REST API.
+ */
+export async function* hasUploadPermissions() {
+	const response = await apiFetch( { path: '/wp/v2/media', method: 'OPTIONS', parse: false } );
```

Comment:
I'm not really clear on why we prefer the approach of making an options request here over that of fetching all the users capabilities from the `/users/me?context=edit` endpoint. The latter would remove the need for all the code that specially handles OPTIONS requests.

@danielbachhuber if you're able to let me know some background behind that decision it would be much appreciated:
https://github.com/WordPress/gutenberg/pull/4155#issuecomment-390029144

Reply:
> I'm not really clear on why we prefer the approach of making an options request here over that of fetching all the users capabilities from the `/users/me?context=edit endpoint`.

See #6361

Reply:
Thanks @danielbachhuber for the info.

That helps me understand why capabilities aren't used across Gutenberg.

Worth noting that for this PR, the permissions check is preloaded server-side during page load and cached. I don't think there's anything that invalidates the cache, the user would have to reload the page if they gained file upload permissions.

Reply:
> I don't think there's anything that invalidates the cache, the user would have to reload the page if they gained file upload permissions.

Yep, this is a reasonable assumption.